Pegasus/include/geometry.hpp

```diff
@@ -0,0 +1,880 @@
+
+        std::array<double, 6> distances;
+        for (uint32_t i = 0; i < distances.size(); ++i) {
+            distances[i] = (cache->aMassCenter - cache->bBoxFaces[i]).magnitude();
+        }
+
+        auto const minIt = std::min_element(distances.begin(), distances.end());
+        auto const minIndex = std::distance(distances.begin(), minIt);
+        cache->contactNormal = cache->bBoxAxes[minIndex].unit();
+
+        return cache->contactNormal;
+    }
+
+    template <>
+    inline double calculatePenetration<Box, Box>(SimpleShape const *a, SimpleShape const *b, CacheBase *cacheBase)
+    {       
+        auto cache = static_cast<Cache<Box, Box>*>(cacheBase);      
+
+        projectAllVertices(cache->contactNormal, cache->aBoxVertices.begin(),
+                           cache->aBoxVertices.end(), cache->aBoxVerticesDistances.begin());
+        projectAllVertices(cache->contactNormal, cache->bBoxVertices.begin(),
+                           cache->bBoxVertices.end(), cache->bBoxVerticesDistances.begin());
+
+        double bMaxVertexDistance = *std::max_element(cache->bBoxVerticesDistances.begin(), cache->bBoxVerticesDistances.end());
+        double aMinVertexDistance = *std::min_element(cache->aBoxVerticesDistances.begin(), cache->aBoxVerticesDistances.end());
+        cache->penetration = bMaxVertexDistance - aMinVertexDistance;
+
+        return cache->penetration;
+    }
+
+} // namespace intersection
+
+    // General intersection
+    using ShapeTypePair = std::pair<SimpleShapeType, SimpleShapeType>;
+
+    size_t shapeTypePairHash(ShapeTypePair const & p);
+
+    class IntersectionQuery {
+    private:
+        std::unordered_map<ShapeTypePair,
+            std::unique_ptr<intersection::CacheBase>,
+            std::function<size_t(ShapeTypePair const & p)> >
+            intersectionCaches;
+
+        std::unordered_map<ShapeTypePair,
+            std::function<void(SimpleShape const *, SimpleShape const *, intersection::CacheBase*)>,
+            std::function<size_t(ShapeTypePair const & p)> >
+            initializeFunctors;
+
+        std::unordered_map<ShapeTypePair,
+            std::function<bool(SimpleShape const *, SimpleShape const *, intersection::CacheBase*)>,
+            std::function<size_t(ShapeTypePair const & p)> >
+            overlapFunctors;
+
+        std::unordered_map<ShapeTypePair,
+            std::function<Vector3(SimpleShape const *, SimpleShape const *, intersection::CacheBase*)>,
+            std::function<size_t(ShapeTypePair const & p)> >
+            calculateContactNormalFunctors;
+
+        std::unordered_map<ShapeTypePair,
+            std::function<double(SimpleShape const *, SimpleShape const *, intersection::CacheBase*)>,
+            std::function<size_t(ShapeTypePair const & p)> >
+            calculatePenetrationFunctors;
+
+    public:
+        IntersectionQuery()
+            : intersectionCaches(11, &shapeTypePairHash)
```

Comment:
It's a prime number for the initial hash table size. I will move it to the constant.